branchpro/posterior.py

```diff
@@ -0,0 +1,199 @@
+# notice and full license details.
+#
+
+import numpy as np
+import scipy.stats
+import pandas as pd
+
+
+class BranchProPosterior(object):
+    """BranchProPosterior Class:
+    Class for computing the posterior distribution used for the inference of
+    the reproduction numbers of an epidemic in the case of a branching process.
+
+    Choice of prior distribution is the conjugate prior for the likelihood
+    (Poisson) of observing given incidence data, hence is a Gamma distribution.
+    We express it in the shape-rate configuration.
+
+    Hence, the posterior distribution will be also be Gamma-distributed.
+
+    Parameters
+    ----------
+    inc_data
+        (pandas Dataframe) contains numbers of new cases by time unit (usually
+        days).
+        Data stored in columns of with one for time and one for incidence
+        number, respectively.
+    daily_serial_interval
+        (list) Unnormalised probability distribution of that the recipient
+        first displays symptoms s days after the infector first displays
+        symptoms.
+    alpha
+        the shape parameter of the Gamma distribution of the prior.
+    beta
+        the rate parameter of the Gamma distribution of the prior.
+    time_key
+        label key given to the temporal data in the inc_data dataframe.
+    inc_key
+        label key given to the incidental data in the inc_data dataframe.
+
+    Notes
+    -----
+    Always apply method run_inference before calling
+    :meth:`BranchProPosterior.get_intervals` to get R behaviour dataframe!
+    """
+
+    def __init__(
+            self, inc_data, daily_serial_interval, alpha, beta,
+            time_key='Time', inc_key='Incidence Number'):
+
+        if not issubclass(type(inc_data), pd.DataFrame):
+            raise TypeError('Incidence data has to be a dataframe')
+
+        try:
+            float(next(iter(daily_serial_interval)))
+        except (TypeError, StopIteration):
+            raise TypeError(
+                'Daily Serial Interval distribution must be iterable')
+        except ValueError:
+            raise TypeError('Daily Serial Interval distribution must contain '
+                            'numeric values')
+
+        if time_key not in inc_data.columns:
+            raise ValueError('No time column with this name in given data')
+
+        if inc_key not in inc_data.columns:
+            raise ValueError(
+                'No incidence column with this name in given data')
+
+        data_times = inc_data[time_key]
+
+        # Pad with zeros the time points where we have no information on
+        # the number of incidences
+        padded_inc_data = inc_data.set_index(time_key).reindex(
+            range(
+                min(data_times), max(data_times)+1)
+                ).fillna(0).reset_index()
+
+        self.cases_data = padded_inc_data[inc_key].to_numpy()
+        self.cases_times = padded_inc_data[time_key]
+        self._serial_interval = daily_serial_interval[::-1]
```

Comment:
Given that we do array multiplication later, it would be a good idea to use np.asarray similar to what you did in models.py